protos/idl/service.proto

```diff
@@ -98,3 +109,68 @@ message Tag {
 message Metadata {
     map<string, string> key_map = 1; // key map is a dictionary of key/val strings that represent metadata
 }
+
+message FilterExpression {
+    repeated SingleValueFilter filters = 1;
+}
+
+message SingleValueFilter {
+    oneof entity {
+        TagFilter tagFilter = 1;
+        PartitionFilter partition_filter = 2;
+        ArtifactFilter artifact_filter = 3;
+        DatasetFilter dataset_filter = 4;
+    }
+
+    // as use-cases come up we can add more operators, ex: gte, like, not eq etc.
+    enum ComparisonOperator {
+        EQUALS = 0;
+    }
+
+    ComparisonOperator operator = 10; // field 10 in case we add more entities to query
+    // Next field number: 11
+}
+
+message ArtifactFilter {
+    oneof property {
+        string artifact_id = 1;
+    }
+}
+
+message TagFilter {
+    oneof property {
+        string tag_name = 1;
+    }
+}
+
+message PartitionFilter {
+    oneof property {
+        KeyValuePair key_val = 1;
+    }
+}
+
+message KeyValuePair {
+    string key = 1;
+    string value = 2;
+}
+
+message DatasetFilter {
+    oneof property {
+        string project = 1;
+        string name = 2;
+        string domain = 3;
+        string version = 4;
+    }
+}
+
+message PaginationOptions {
+    int32 limit = 1;
```

Comment:
Any specific reason to go with offset and limit vs limit and last token?

Reply:
Yea I think token would be more flexible. V1 used offset so at first I think we will want to return offset as the `token` and then later we can transition to a token that is more robust like `created_at`.